### ISSUE TITLE
Remove hiera_hash in favour of lookup()

### ISSUE THREAD

**Issue by brettswift** [elastic/puppet-elasticsearch#642]
Any hiera functions prevent you from moving to the new puppet lookup() function.

When moving to the hierarchy in modules,  you have to turn off 'classic hiera'. 

Classic hiera uses:  hiera(),  hiera_array(),  hiera_hash().

The new version uses lookup().  

This would change 

`hiera_hash('elasticsearch::plugins', $::elasticsearch::plugins)` 
to: 
```
lookup( { 'name'  => 'elasticsearch::plugins',
               'merge' => 'hash',
        })
```

Happy to submit a PR on this - unless anyone knows that this is in the works and I didn't see it when searching issues.

**Comment by tylerjl** [comment]
Hey @brettswift, I don't think this is underway in any issues or PRs, so feel free to take a crack at it if you'd like. I haven't played around with `lookup()` vs. `hiera*()` before, but as long as it's supported across the versions the module supports and tests pass, I'm  👍. Thanks for the help!

**Comment by brettswift** [comment]
It seems like this was introduced into puppet 4.3.

According to the metadata.json file here,  the module still supports 3.2 on PE. (which I think was puppet 3.6 or something). 

Is there a version / timeline to leave puppet 3 behind?

**Comment by tylerjl** [comment]
No timeline right now. In the meantime one option could be handling lookup()/hiera() calls based upon the puppet version fact - it's definitely uglier than just deprecating 3.x but it'd be immediately doable.

**Comment by brettswift** [comment]
Good idea,   I thought the compiler would see the function and fail if it wasn't available, even if it wasn't going to execute it in that run.. however a quick test proves me wrong!

I am currently doing inventory of all the things we'd have to refactor to get to the next version of hiera.. so you may not see this PR immediately.   I'll see what I can do.   Thanks!

**Comment by brettswift** [comment]
There is a nicer workaround here: 

https://tickets.puppetlabs.com/browse/HI-518

If that doesn't work I'll send in a PR to use the new lookup function if it's available.    

I'll leave the state of this ticket up to you if you want to close it or keep it for tracking etc..

**Comment by tylerjl** [comment]
If the workaround ends up working for you and you're happy with the solution, feel free to close this issue. If you need to PR something, go ahead and we can close this out once that's done. 👍

**Comment by brettswift** [comment]
Will do. Thanks

**Comment by tylerjl** [comment]
FYI, I'll be removing all calls to `hiera_*` functions in the module when the 6.x branch of the module gets release and we deprecate Puppet 3.x. Moving forward, folks should be able to use [`lookup_options`](https://docs.puppet.com/puppet/5.0/hiera_merging.html#configuring-merge-behavior-in-hiera-data) instead and get the same effect without needing any calls to `lookup()`.
